File: src/types/account.ts
export type AccountType =
  | "ROOT"
  | "ASSET"
  | "BANK"
  | "CASH"
  | "INVESTMENT"
  | "STOCK"
  | "MUTUAL_FUND"
  | "RECEIVABLE"
  | "LIABILITY"
  | "CREDIT_CARD"
  | "PAYABLE"
  | "EQUITY"
  | "INCOME"
  | "EXPENSE"

export interface Account {
  id: number
  parentId: number | null
  accountType: AccountType
  code: string | null
  name: string
  description: string | null
  currencyCode: string
  isPlaceholder: boolean
  isActive: boolean
  sortOrder: number
  balance: string
  createdAt: string
  updatedAt: string
}

export interface AccountNode {
  id: number
  parentId: number | null
  accountType: AccountType
  code: string | null
  name: string
  description: string | null
  currencyCode: string
  isPlaceholder: boolean
  isActive: boolean
  sortOrder: number
  balance: string
  children?: AccountNode[]
}

export interface AccountBalance {
  accountId: number
  balance: string
}

export const ACCOUNT_TYPE_LABELS: Record<AccountType, string> = {
  ROOT: "Root",
  ASSET: "Asset",
  BANK: "Bank",
  CASH: "Cash",
  INVESTMENT: "Investment",
  STOCK: "Stock",
  MUTUAL_FUND: "Mutual Fund",
  RECEIVABLE: "Accounts Receivable",
  LIABILITY: "Liability",
  CREDIT_CARD: "Credit Card",
  PAYABLE: "Accounts Payable",
  EQUITY: "Equity",
  INCOME: "Income",
  EXPENSE: "Expense",
}

export const ACCOUNT_TYPE_TAXONOMY: AccountType[] = [
  "ROOT",
  "ASSET",
  "BANK",
  "CASH",
  "INVESTMENT",
  "STOCK",
  "MUTUAL_FUND",
  "RECEIVABLE",
  "LIABILITY",
  "CREDIT_CARD",
  "PAYABLE",
  "EQUITY",
  "INCOME",
  "EXPENSE",
]

export function isAccountParentType(t: AccountType): boolean {
  return ["ROOT", "ASSET", "LIABILITY", "EQUITY", "INCOME", "EXPENSE"].includes(t)
}

export function getAccountNormalBalance(t: AccountType): "debit" | "credit" {
  switch (t) {
    case "ASSET":
    case "BANK":
    case "CASH":
    case "INVESTMENT":
    case "STOCK":
    case "MUTUAL_FUND":
    case "RECEIVABLE":
    case "EXPENSE":
      return "debit"
    case "LIABILITY":
    case "CREDIT_CARD":
    case "PAYABLE":
    case "EQUITY":
    case "INCOME":
    case "ROOT":
      return "credit"
  }
}

File: src-tauri/src/models/account.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Account {
    pub id: i64,
    pub parent_id: Option<i64>,
    pub account_type: String,
    pub code: Option<String>,
    pub name: String,
    pub description: Option<String>,
    pub currency_code: String,
    pub is_placeholder: bool,
    pub is_active: bool,
    pub sort_order: i64,
    pub balance: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct AccountNode {
    pub id: i64,
    pub parent_id: Option<i64>,
    pub account_type: String,
    pub code: Option<String>,
    pub name: String,
    pub description: Option<String>,
    pub currency_code: String,
    pub is_placeholder: bool,
    pub is_active: bool,
    pub sort_order: i64,
    pub balance: String,
    pub children: Vec<AccountNode>,
}

#[derive(Debug, Deserialize)]
pub struct CreateAccountPayload {
    pub parent_id: Option<i64>,
    pub account_type: String,
    pub code: Option<String>,
    pub name: String,
    pub description: Option<String>,
    pub currency_code: Option<String>,
    pub is_placeholder: Option<bool>,
    pub sort_order: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateAccountPayload {
    pub id: i64,
    pub parent_id: Option<Option<i64>>,
    pub account_type: Option<String>,
    pub code: Option<Option<String>>,
    pub name: Option<String>,
    pub description: Option<Option<String>>,
    pub currency_code: Option<String>,
    pub is_placeholder: Option<bool>,
    pub is_active: Option<bool>,
    pub sort_order: Option<i64>,
}
